File: reference/js/garlicstamp.mjs
// GarlicStamp v0.6 JavaScript reference verifier for Node.js 20+.
//
// Validates Alpha Garage-issued GarlicStamp portable credentials without private
// Garage context: fetch the public key, verify the Ed25519 signature over
// canonical JSON, then check the portable v0.6 evidence fields.

import { createPublicKey, verify as cryptoVerify } from 'node:crypto';

export const SUPPORTED_VERSION = '0.6';
export const DEFAULT_BASE_URL = 'https://alphagarage.io';
export const DEFAULT_USER_AGENT = 'garlicstamp-js-reference/0.6';

class JsonNumber {
  constructor(raw) {
    this.raw = raw;
    this.value = Number(raw);
  }

  valueOf() { return this.value; }
  toJSON() { return this.value; }
}

function isJsonNumber(value) {
  return value instanceof JsonNumber;
}

function isPlainObject(value) {
  return value !== null && typeof value === 'object' && !Array.isArray(value) && !isJsonNumber(value);
}

export function parseEnvelopeJson(text) {
  // Tiny lossless JSON parser: preserves number lexemes such as 0.0 because
  // Alpha Garage v0.6 signs Python json.dumps output, where 0 and 0.0 are not
  // interchangeable at the byte level. JSON.parse helpfully erases that detail.
  let i = 0;
  const s = String(text);
  const ws = () => { while (/\s/.test(s[i] ?? '')) i += 1; };
  const expect = (token) => {
    if (!s.startsWith(token, i)) throw new SyntaxError(`expected ${token} at ${i}`);
    i += token.length;
  };
  const parseString = () => {
    const start = i;
    i += 1;
    while (i < s.length) {
      const ch = s[i++];
      if (ch === '\\') i += 1;
      else if (ch === '"') return JSON.parse(s.slice(start, i));
    }
    throw new SyntaxError('unterminated JSON string');
  };
  const parseNumber = () => {
    const start = i;
    if (s[i] === '-') i += 1;
    if (s[i] === '0') i += 1;
    else if (/[1-9]/.test(s[i] ?? '')) while (/[0-9]/.test(s[i] ?? '')) i += 1;
    else throw new SyntaxError(`bad JSON number at ${i}`);
    if (s[i] === '.') { i += 1; if (!/[0-9]/.test(s[i] ?? '')) throw new SyntaxError(`bad JSON number at ${i}`); while (/[0-9]/.test(s[i] ?? '')) i += 1; }
    if (s[i] === 'e' || s[i] === 'E') { i += 1; if (s[i] === '+' || s[i] === '-') i += 1; if (!/[0-9]/.test(s[i] ?? '')) throw new SyntaxError(`bad JSON number at ${i}`); while (/[0-9]/.test(s[i] ?? '')) i += 1; }
    return new JsonNumber(s.slice(start, i));
  };
  const parseArray = () => {
    i += 1; ws();
    const out = [];
    if (s[i] === ']') { i += 1; return out; }
    while (true) {
      out.push(parseValue()); ws();
      if (s[i] === ']') { i += 1; return out; }
      expect(','); ws();
    }
  };
  const parseObject = () => {
    i += 1; ws();
    const out = {};
    if (s[i] === '}') { i += 1; return out; }
    while (true) {
      const key = parseString(); ws(); expect(':'); ws();
      out[key] = parseValue(); ws();
      if (s[i] === '}') { i += 1; return out; }
      expect(','); ws();
    }
  };
  const parseValue = () => {
    ws();
    const ch = s[i];
    if (ch === '"') return parseString();
    if (ch === '{') return parseObject();
    if (ch === '[') return parseArray();
    if (ch === 't') { expect('true'); return true; }
    if (ch === 'f') { expect('false'); return false; }
    if (ch === 'n') { expect('null'); return null; }
    return parseNumber();
  };
  const value = parseValue(); ws();
  if (i !== s.length) throw new SyntaxError(`unexpected JSON trailing content at ${i}`);
  return value;
}

export function canonicalJson(value) {
  if (isJsonNumber(value)) return value.raw;
  if (Array.isArray(value)) {
    return `[${value.map((item) => canonicalJson(item)).join(', ')}]`;
  }
  if (isPlainObject(value)) {
    return `{${Object.keys(value).sort().map((key) => `${JSON.stringify(key)}: ${canonicalJson(value[key])}`).join(', ')}}`;
  }
  return JSON.stringify(value);
}

export function dottedGet(payload, path) {
  let cur = payload;
  for (const part of path.split('.')) {
    if (!isPlainObject(cur) || !(part in cur)) return undefined;
    cur = cur[part];
  }
  return cur;
}

export function missingPortableFields(credential) {
  const required = [
    'protocol',
    'version',
    'issuer.id',
    'subject.id',
    'subject.type',
    'claims.verification_sources',
    'claims.performance',
  ];
  const missing = [];
  for (const path of required) {
    const value = dottedGet(credential, path);
    if (value === undefined || value === null || value === '' || (Array.isArray(value) && value.length === 0) || (isPlainObject(value) && Object.keys(value).length === 0)) {
      missing.push(path);
    }
  }

  const sources = dottedGet(credential, 'claims.verification_sources');
  if (Array.isArray(sources)) {
    sources.forEach((source, index) => {
      if (!isPlainObject(source)) {
        missing.push(`claims.verification_sources[${index}]`);
        return;
      }
      if (!source.type) missing.push(`claims.verification_sources[${index}].type`);
      if (!isPlainObject(source.issuer) || !source.issuer.id) missing.push(`claims.verification_sources[${index}].issuer.id`);
      if (!source.evidence_url) missing.push(`claims.verification_sources[${index}].evidence_url`);
    });
  }

  const performance = dottedGet(credential, 'claims.performance');
  if (isPlainObject(performance)) {
    if (!isPlainObject(performance.source) || !performance.source.id) missing.push('claims.performance.source.id');
    if (!performance.evidence_url) missing.push('claims.performance.evidence_url');
    if (!isPlainObject(performance.windows) || !isPlainObject(performance.windows.all_time)) missing.push('claims.performance.windows.all_time');
  }

  return [...new Set(missing)].sort();
}

export async function fetchCredential(agentIdOrSlug, { baseUrl = DEFAULT_BASE_URL } = {}) {
  const response = await fetch(`${baseUrl.replace(/\/$/, '')}/api/garage/verify/${encodeURIComponent(agentIdOrSlug)}`, {
    headers: { accept: 'application/json', 'user-agent': DEFAULT_USER_AGENT },
  });
  if (!response.ok) throw new Error(`credential fetch failed: ${response.status}`);
  return parseEnvelopeJson(await response.text());
}

export async function fetchPublicKey({ baseUrl = DEFAULT_BASE_URL } = {}) {
  const response = await fetch(`${baseUrl.replace(/\/$/, '')}/api/garage/garlicstamp-pubkey`, {
    headers: { accept: 'application/json', 'user-agent': DEFAULT_USER_AGENT },
  });
  if (!response.ok) throw new Error(`public key fetch failed: ${response.status}`);
  const body = await response.json();
  if (body.algorithm !== 'Ed25519') throw new Error(`unsupported GarlicStamp key algorithm: ${body.algorithm}`);
  const raw = Buffer.from(body.public_key, 'base64');
  if (raw.length !== 32) throw new Error('GarlicStamp Ed25519 public key must decode to 32 bytes');
  return raw;
}

function decodeSignature(signature) {
  if (typeof signature !== 'string') return null;
  const raw = Buffer.from(signature, 'base64');
  if (raw.length !== 64 || raw.toString('base64') !== signature) return null;
  return raw;
}

function ed25519PublicKeyFromRaw(rawPublicKey) {
  // RFC 8410 SubjectPublicKeyInfo prefix for a raw Ed25519 public key.
  const spkiPrefix = Buffer.from('302a300506032b6570032100', 'hex');
  return createPublicKey({ key: Buffer.concat([spkiPrefix, rawPublicKey]), format: 'der', type: 'spki' });
}

export function verifySignature(credential, signature, publicKey) {
  const signatureBytes = decodeSignature(signature);
  if (!signatureBytes) return false;
  return cryptoVerify(null, Buffer.from(canonicalJson(credential), 'utf8'), ed25519PublicKeyFromRaw(publicKey), signatureBytes);
}

export async function verify(envelope, { baseUrl = DEFAULT_BASE_URL, publicKey = null } = {}) {
  const credential = isPlainObject(envelope) ? envelope.credential : null;
  const signature = isPlainObject(envelope) ? envelope.signature : null;
  if (!isPlainObject(credential) || typeof signature !== 'string') {
    return {
      valid: false,
      subject_id: null,
      checks: { signature: false, schema: false },
      reason: 'missing_credential_or_signature',
      error_code: 'missing_credential_or_signature',
      missing: ['credential', 'signature'],
    };
  }

  const subjectId = dottedGet(credential, 'subject.id') ?? null;
  let missing = missingPortableFields(credential);
  let schemaReason = null;
  let schemaOk = missing.length === 0;
  if (credential.version !== SUPPORTED_VERSION) {
    missing = [...new Set([...missing, 'version'])].sort();
    schemaOk = false;
    schemaReason = 'unsupported_version';
  } else if (!schemaOk) {
    schemaReason = 'missing_required_fields';
  }

  if (!decodeSignature(signature)) {
    return {
      valid: false,
      subject_id: subjectId,
      checks: { signature: false, schema: schemaOk },
      reason: 'malformed_signature',
      error_code: 'malformed_signature',
      missing,
    };
  }

  const key = publicKey ?? await fetchPublicKey({ baseUrl });
  const signatureOk = verifySignature(credential, signature, key);
  const valid = signatureOk && schemaOk;
  const reason = valid ? null : (schemaReason ?? 'signature_mismatch');
  return {
    valid,
    subject_id: subjectId,
    checks: { signature: signatureOk, schema: schemaOk },
    reason,
    error_code: reason,
    missing,
  };
}

File: reference/python/garlicstamp.py
"""GarlicStamp v0.6 Python reference verifier.

This file is intentionally small and dependency-light so third-party platforms can
copy it into an integration test or vendor it directly. It validates Alpha
Garage-issued GarlicStamp portable credentials without private Garage context:
fetch the public key, verify the Ed25519 signature over canonical JSON, then
check the portable v0.6 evidence fields.

Install dependency:
    python -m pip install cryptography requests
"""

from __future__ import annotations

import base64
import json
from typing import Any

import requests
from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey

SUPPORTED_VERSION = "0.6"
DEFAULT_BASE_URL = "https://alphagarage.io"
DEFAULT_USER_AGENT = "garlicstamp-python-reference/0.6"


class VerificationResult:
    def __init__(
        self,
        *,
        valid: bool,
        subject_id: str | None,
        checks: dict[str, bool],
        reason: str | None = None,
        error_code: str | None = None,
        missing: list[str] | None = None,
    ) -> None:
        self.valid = valid
        self.subject_id = subject_id
        self.checks = checks
        self.reason = reason
        self.error_code = error_code or reason
        self.missing = missing or []

    def to_dict(self) -> dict[str, Any]:
        return {
            "valid": self.valid,
            "subject_id": self.subject_id,
            "checks": self.checks,
            "reason": self.reason,
            "error_code": self.error_code,
            "missing": self.missing,
        }


def canonical_json(payload: dict[str, Any]) -> bytes:
    """Return the exact bytes Alpha Garage signs for a credential object."""
    return json.dumps(payload, sort_keys=True, default=str).encode("utf-8")


def fetch_credential(agent_id_or_slug: str, *, base_url: str = DEFAULT_BASE_URL) -> dict[str, Any]:
    response = requests.get(
        f"{base_url.rstrip('/')}/api/garage/verify/{agent_id_or_slug}",
        headers={"accept": "application/json", "user-agent": DEFAULT_USER_AGENT},
        timeout=30,
    )
    response.raise_for_status()
    return response.json()


def fetch_public_key(*, base_url: str = DEFAULT_BASE_URL) -> bytes:
    response = requests.get(
        f"{base_url.rstrip('/')}/api/garage/garlicstamp-pubkey",
        headers={"accept": "application/json", "user-agent": DEFAULT_USER_AGENT},
        timeout=30,
    )
    response.raise_for_status()
    body = response.json()
    if body.get("algorithm") != "Ed25519":
        raise ValueError(f"unsupported GarlicStamp key algorithm: {body.get('algorithm')!r}")
    raw_key = base64.b64decode(body["public_key"], validate=True)
    if len(raw_key) != 32:
        raise ValueError("GarlicStamp Ed25519 public key must decode to 32 bytes")
    return raw_key


def _dotted_get(payload: dict[str, Any], path: str) -> Any:
    cur: Any = payload
    for part in path.split("."):
        if not isinstance(cur, dict) or part not in cur:
            return None
        cur = cur[part]
    return cur


def missing_portable_fields(credential: dict[str, Any]) -> list[str]:
    """Return required v0.6 fields missing from the credential.

    GarlicStamp is agent-first: no human owner field is required. The required
    claims are issuer-observed proof bundles, not self-attested vanity claims.
    """
    required = [
        "protocol",
        "version",
        "issuer.id",
        "subject.id",
        "subject.type",
        "claims.verification_sources",
        "claims.performance",
    ]
    missing: list[str] = []
    for path in required:
        value = _dotted_get(credential, path)
        if value in (None, "", [], {}):
            missing.append(path)

    sources = _dotted_get(credential, "claims.verification_sources")
    if isinstance(sources, list):
        for index, source in enumerate(sources):
            if not isinstance(source, dict):
                missing.append(f"claims.verification_sources[{index}]")
                continue
            if not source.get("type"):
                missing.append(f"claims.verification_sources[{index}].type")
            issuer = source.get("issuer")
            if not isinstance(issuer, dict) or not issuer.get("id"):
                missing.append(f"claims.verification_sources[{index}].issuer.id")
            if not source.get("evidence_url"):
                missing.append(f"claims.verification_sources[{index}].evidence_url")

    performance = _dotted_get(credential, "claims.performance")
    if isinstance(performance, dict):
        source = performance.get("source")
        if not isinstance(source, dict) or not source.get("id"):
            missing.append("claims.performance.source.id")
        if not performance.get("evidence_url"):
            missing.append("claims.performance.evidence_url")
        windows = performance.get("windows")
        if not isinstance(windows, dict) or not isinstance(windows.get("all_time"), dict):
            missing.append("claims.performance.windows.all_time")

    return sorted(set(missing))


def _decode_signature(signature: str) -> bytes | None:
    if not isinstance(signature, str):
        return None
    try:
        raw = base64.b64decode(signature, validate=True)
    except Exception:
        return None
    return raw if len(raw) == 64 else None


def verify_signature(credential: dict[str, Any], signature: str, public_key: bytes) -> bool:
    signature_bytes = _decode_signature(signature)
    if signature_bytes is None:
        return False
    try:
        Ed25519PublicKey.from_public_bytes(public_key).verify(signature_bytes, canonical_json(credential))
        return True
    except InvalidSignature:
        return False


def verify(envelope: dict[str, Any], *, base_url: str = DEFAULT_BASE_URL, public_key: bytes | None = None) -> VerificationResult:
    """Verify a GarlicStamp envelope and return explicit pass/fail checks."""
    credential = envelope.get("credential") if isinstance(envelope, dict) else None
    signature = envelope.get("signature") if isinstance(envelope, dict) else None
    if not isinstance(credential, dict) or not isinstance(signature, str):
        return VerificationResult(
            valid=False,
            subject_id=None,
            checks={"signature": False, "schema": False},
            reason="missing_credential_or_signature",
            missing=["credential", "signature"],
        )

    subject_id = _dotted_get(credential, "subject.id")
    missing = missing_portable_fields(credential)
    if credential.get("version") not in (SUPPORTED_VERSION,):
        missing = sorted(set(missing + ["version"]))
        schema_ok = False
        schema_reason = "unsupported_version"
    else:
        schema_ok = not missing
        schema_reason = "missing_required_fields" if missing else None

    if _decode_signature(signature) is None:
        return VerificationResult(
            valid=False,
            subject_id=subject_id,
            checks={"signature": False, "schema": schema_ok},
            reason="malformed_signature",
            missing=missing,
        )

    key = public_key if public_key is not None else fetch_public_key(base_url=base_url)
    signature_ok = verify_signature(credential, signature, key)
    valid = signature_ok and schema_ok
    reason = None if valid else (schema_reason or "signature_mismatch")
    return VerificationResult(
        valid=valid,
        subject_id=subject_id,
        checks={"signature": signature_ok, "schema": schema_ok},
        reason=reason,
        missing=missing,
    )
